### DISCUSSION TITLE
Dispose of a resource held by a cached query

### DISCUSSION THREAD

**Discussion by danielbrauer** [TanStack/query#1109] [General]
Is there any way to have an action performed when a query result is overwritten or removed from the cache? I would like to have my query return URLs created with `URL.createObjectURL()`but currently can't see how I would issue the corresponding `URL.revokeObjectURL()` to avoid leaking resources.

**Comment by TkDodo** [comment]
That's an interesting question. As far as I am aware, there is no dedicated callback that will be invoked when a query is evicted from the cache, but you can [globally subscribe](https://react-query.tanstack.com/docs/api/#querycachesubscribe) to the query cache and act on the specific removal there

**Reply by danielbrauer** [reply]
Thanks for the idea. The docs on query cache subscription are a bit lean, so I wasn't sure if the events would be appropriate. I'll try it out, though!

Do you know if QueryObserver might be better for this thing once it's expose? https://github.com/tannerlinsley/react-query/commit/5d0546426ca90775ea38314e0110537afdddeaaa

**Reply by TkDodo** [reply]
It looks like `queryCache.watchQuery` was just released a couple of days ago in [v2.23.0](https://github.com/tannerlinsley/react-query/releases/tag/v2.23.0). There are no docs for it yet, so you'd have to try what that does :)

**Comment by edzis** [comment]
@danielbrauer How did you end up solving this? I have the same need.

**Comment by TkDodo** [comment]
Isn't the problem also that if the queryFn is invoked because of a background refetch, you would create a new resource that overrides the old one without releasing it? So this doesn't necessarily have something to do with cache garbage collection, unless you combine it with infinite staleTime...

**Comment by edzis** [comment]
I ended up using this - https://gist.github.com/edzis/e97d83ee4eb7bf016f5dff165244ebd5

```
const queryClient = new QueryClient();
queryClient.getQueryCache().subscribe((event) => {
  if (!event) {
    return;
  }
  if (event.type === 'queryRemoved') {
    maybeRevokeBlobImageUrl(event.query);
  }
});
```

```


export const maybeRevokeBlobImageUrl = (query: Query) => {
  if (
    typeof query.state.data !== 'string' ||
    query.queryKey[0] !== QUERY_KEY_BASE
  ) {
    return;
  }

  const url = query.state.data;
  if (DEBUG) {
    console.log('BlobImage revokeObjectURL', url);
  }
  URL.revokeObjectURL(url);
};
```

**Comment by TkDodo** [comment]
this wouldn't work in the following case:

```
const { data, refetch } = useQuery('key', () => {
  return URL.createObjectURL(...)
})

<button onClick={() => refetch()}>refetch</button>
```

Every time you refetch a query, the `queryFn` is executed. Even focussing the window is enough to trigger a background refetch, which executes the queryFn. At no time in between will you get a `queryRemoved` event, because the query is not removed.

**Reply by edzis** [reply]
But would there be any danger in my case where I keep the query inside the component without a way to refetch it and make it infinitely stale?

**Reply by TkDodo** [reply]
if you ensure that the queryFn never runs again then yes, there is no problem. I would still generally try to keep these things outside of react-query, e.g. by coupling the creation to the component lifecycle with `useEffect` because you will have great cleanup options.

**Reply by edzis** [reply]
I recognize this is tangling the systems and your suggestion would be better for simple scenarios.
But I chose to use react-query to better support my usecase - the same url must be shown in different image instances over time. Creating and deleting blob urls locally in the component would result in duplicates and/or caching the full image data in react-query. And that nudges to reinvent react-query on the side for tracking the number of active uses of a blob url to know when an url can be released.

**Reply by TkDodo** [reply]
The only way I could see this working is by returning a cleanup function from the `queryFn`, which we could run before the `queryFn` is run the next time. However, we can't do that because the `queryFn` returns a `Promise` already.

Also, the `queryFn` should be idempotent, as it can be executed at arbitrary times (smart background refetches), and it can also be cancelled. It shouldn't have any observable side-effects like creating objects that need a cleanup. It's the same principle why the resource on the backend (GET endpoint) shouldn't have any observable side-effects.

**Comment by wolever** [comment]
Here's a little helper which will call `Symbol.dispose` on values as they are removed from the QueryCache: https://gist.github.com/wolever/30a34443a9b2982615bbd2e84cd9529e

```js
/**
 * Subscribe to ``client``'s query cache and dispose of any values that have a
 * ``[Symbol.dispose]`` method.
 *
 * Notes:
 * - The exported ``SymbolDispose`` can be used if ``Symbol.dispose`` is not
 *   available.
 * - The ``createDisposingObjectUrl`` helper can be used to create a disposable
 *   object URL.
 *
 * For example::
 *
 *   const client = new QueryClient()
 *   queryClientDisposeOnCacheEvict(client)
 *
 *   export const MyComponent = () => {
 *     const imageQuery = useQuery({
 *       queryFn: () => {
 *         const blob = await fetch(...).then(r => r.blob())
 *         // Note: see the ``createDisposingObjectUrl`` helper.
 *         const url = URL.createObjectURL(blob)
 *         return { url, [Symbol.dispose]: () => URL.revokeObjectURL(url) }
 *       },
 *     })
 *     ...
 *   }
 */
```

**Reply by TkDodo** [reply]
this looks really promising, thanks for sharing. I thought that you also need to remove when you receive an `update` event, but you are doing that inside of `add` 👍

**Reply by TkDodo** [reply]
if you want, we can add this to [community projects](https://tanstack.com/query/v5/docs/react/community/community-projects) ?

**Reply by spaguette** [reply]
I have used your suggestion and it works amazingly, however I had to turn off the `structuralSharing` in the `queryOptions` (otherwise after the first invalidation it will stop passing the `Symbol.disposer)

I.e.
```
useQuery({
    queryKey: myQueryKey,
    queryFn: async () => {
      const blob = await getBlob();
      return createDisposingObjectUrl(blob);
    },
    // NB! VERY important to set the structural sharing to false for this.
    // Otherwise the disposal will stop working (the symbol is then not saved in the query state on invalidation)
    structuralSharing: false,
  });
```

Which also led me to a thought: is it possible to actually handle the object revoking in the structuralSharing callback since it passes both the old and the new value? :thinking:

The reason I'm asking is that it can be a common case to fetch an auth-protected files / images (like the ones that are returned by presigned AWS S3 links). Since they don't auto-renew, even though they support caching via Etag, it's easier to just cache it on the client, e.g. with React Query.

**Comment by jonahallibone** [comment]
I am kind of curious if there is any better methods for this yet? I feel like it's a relatively valuable addition to the core library.
